### ISSUE TITLE
Breaking change in electron-updater 4.3.10

### ISSUE THREAD

**Issue by davej** [electron-userland/electron-builder#6000]
* **Version**: 4.3.10

* **Electron Version**: Effects Electron 11 and before. Does not effect Electron 12 or later.

* **Target**: All

Electron-updater has introduced `fs/promises` which will break the build on Electron 11 and earlier. This is because `fs/promises` is not supported in Node 12  (which is used by Electron 11). The change was introduced in this PR: https://github.com/electron-userland/electron-builder/pull/5890

This is a critical issue for anybody on Electron 11 or earlier because it will break their application after an update and it is not possible to recover by pushing a fix in an update.

The best mitigation that I think of here is to publish a `4.3.11` that does not use `fs/promises`. If you want to use `fs/promises` then you can publish `5.0.0` that uses `fs/promises`.

![electron-updater](https://user-images.githubusercontent.com/26971/123619180-d6c89000-d800-11eb-964b-2fab0aa8eeb3.png)

**Comment by shownoso** [comment]
I got the same problem

**Comment by tub5** [comment]
This has halted all development on our apps

**Comment by theolivenbaum** [comment]
Same here

**Comment by sFireFrog** [comment]
。。。I solved this problem by specifying the version in package.json

**Comment by tub5** [comment]
> 。。。I solved this problem by specifying the version in package.json

That would be a temporary fix.
However, we're using the [Electron.NET](https://github.com/ElectronNET/Electron.NET) package which uses a CLI to build our apps. Therefore we can't just change it.
With a breaking change such as this, it shouldn't have been placed as a minor version.

**Comment by sFireFrog** [comment]
> > 。。。I solved this problem by specifying the version in package.json
> 
> That would be a temporary fix.
> However, we're using the [Electron.NET](https://github.com/ElectronNET/Electron.NET) package which uses a CLI to build our apps. Therefore we can't just change it.
> With a breaking change such as this, it shouldn't have been placed as a minor version.

Yes, I also use Electron.NET, this problem has troubled me for a long time, I directly modified the CLI

**Comment by tub5** [comment]
> > > 。。。I solved this problem by specifying the version in package.json
> > 
> > 
> > That would be a temporary fix.
> > However, we're using the [Electron.NET](https://github.com/ElectronNET/Electron.NET) package which uses a CLI to build our apps. Therefore we can't just change it.
> > With a breaking change such as this, it shouldn't have been placed as a minor version.
> 
> Yes, I also use Electron.NET, this problem has troubled me for a long time, I directly modified the CLI

Nice. [I have put a PR in](https://github.com/ElectronNET/Electron.NET/pull/579) to upgrade us to Electron 12.0.12 and Tweeted Gregor. No PR's since Jan makes for a sad dev team.

**Comment by davej** [comment]
> I solved this problem by specifying the version in package.json

Most users won't be aware of this issue before they publish. For this reason, it's best to fix this in a patch release.

**Comment by ekuchynski** [comment]
Also got this surprise today. Built and deployed updates refused to start on many user machines.. :( Please help.

**Comment by theolivenbaum** [comment]
@davej: I'm publishing a temp nuget package with the fixes your PR (and including also the fix for the splashscreen that has been long pending). It's based on my fork here: https://github.com/theolivenbaum/Electron.NET

If you want to use it, you'll need to:
- change the NuGet package: https://www.nuget.org/packages/h5.ElectronNET.API/
- uninstall the electronize CLI and install this version instead: https://www.nuget.org/packages/h5.ElectronNET.CLI/
- change any calls to the CLI from ``electronize`` to ``electronize-h5``


As this is now using Electron12, you also need to add this to your ``CreateWindow`` calls:

````` csharp
Electron.WindowManager.CreateWindowAsync(new BrowserWindowOptions
{
   // ...
    WebPreferences = new WebPreferences()
    {
        NodeIntegration = true,
        ContextIsolation = false
    },
   // ...
``````

We're using this one till it's fixed upstream, and testing if everything still works. If you want to use it for now and hit anything, drop me an email to ``rafael (at) curiosity (dot) ai``

**Comment by skelt** [comment]
we need to use nodejs 12 for windows server 2003.  is there a solution that works with nodejs12?

**Comment by davej** [comment]
@develar @mmaietta I usually don't like to tag core devs but I think this is particularly nasty given that there isn't a way to recover once a problematic version has been shipped (apart from emailing affected end-users to download again).

**Comment by develar** [comment]
`4.3.9` is marked as `latest` again for now.

`4.3.10` was marked as a `next` for a two months.

**Comment by mankeheaven** [comment]
hello, which version can I rollback to differential download?  
link #5948
Is there anyway to differential download lower than 4.3.9, other than fork this resp?

**Comment by skelt** [comment]
electronize build works using nodejs14 and with 4.3.9 set to the latest we're up and running again on win server 2003

**Comment by tub5** [comment]
> @davej: I'm publishing a temp nuget package with the fixes your PR (and including also the fix for the splashscreen that has been long pending). It's based on my fork here: https://github.com/theolivenbaum/Electron.NET
> 
> If you want to use it, you'll need to:
> 
> * change the NuGet package: https://www.nuget.org/packages/h5.ElectronNET.API/
> * uninstall the electronize CLI and install this version instead: https://www.nuget.org/packages/h5.ElectronNET.CLI/
> * change any calls to the CLI from `electronize` to `electronize-h5`
> 
> As this is now using Electron12, you also need to add this to your `CreateWindow` calls:
> 
> ```cs
> Electron.WindowManager.CreateWindowAsync(new BrowserWindowOptions
> {
>    // ...
>     WebPreferences = new WebPreferences()
>     {
>         NodeIntegration = true,
>         ContextIsolation = false
>     },
>    // ...
> ```
> 
> We're using this one till it's fixed upstream, and testing if everything still works. If you want to use it for now and hit anything, drop me an email to `rafael (at) curiosity (dot) ai`

Cheers for that.
Gregor is reviewing the PR's Thursday evening so hopefully we'll have an update shortly after.

**Comment by Mgonand** [comment]
For a quick workaround, you can patch (using [patch-package](https://www.npmjs.com/package/patch-package) npm if you want to persist through commits) the AppUpdater.js file (located in node_modules/electron-updater/out/AppUpdater.js):

```diff
diff --git a/node_modules/electron-updater/out/AppUpdater.js b/node_modules/electron-updater/out/AppUpdater.js
index 42ed9a3..469fa39 100644
--- a/node_modules/electron-updater/out/AppUpdater.js
+++ b/node_modules/electron-updater/out/AppUpdater.js
@@ -5,7 +5,6 @@ const builder_util_runtime_1 = require("builder-util-runtime");
 const crypto_1 = require("crypto");
 const events_1 = require("events");
 const fs_extra_1 = require("fs-extra");
-const promises_1 = require("fs/promises");
 const js_yaml_1 = require("js-yaml");
 const lazy_val_1 = require("lazy-val");
 const path = require("path");
@@ -377,7 +376,7 @@ class AppUpdater extends events_1.EventEmitter {
         if (this._appUpdateConfigPath == null) {
             this._appUpdateConfigPath = this.app.appUpdateConfigPath;
         }
-        return js_yaml_1.load(await promises_1.readFile(this._appUpdateConfigPath, "utf-8"));
+        return js_yaml_1.load(await fs_extra_1.readFile(this._appUpdateConfigPath, "utf-8"));
     }
     computeRequestHeaders(provider) {
         const fileExtraDownloadHeaders = provider.fileExtraDownloadHeaders;
@@ -395,7 +394,7 @@ class AppUpdater extends events_1.EventEmitter {
     async getOrCreateStagingUserId() {
         const file = path.join(this.app.userDataPath, ".updaterId");
         try {
-            const id = await promises_1.readFile(file, "utf-8");
+            const id = await fs_extra_1.readFile(file, "utf-8");
             if (builder_util_runtime_1.UUID.check(id)) {
                 return id;
             }
@@ -477,7 +476,7 @@ class AppUpdater extends events_1.EventEmitter {
         }
         const downloadedUpdateHelper = await this.getOrCreateDownloadHelper();
         const cacheDir = downloadedUpdateHelper.cacheDirForPendingUpdate;
-        await promises_1.mkdir(cacheDir, { recursive: true });
+        await fs_extra_1.mkdir(cacheDir, { recursive: true });
         const updateFileName = getCacheUpdateFileName();
         let updateFile = path.join(cacheDir, updateFileName);
         const packageFile = packageInfo == null ? null : path.join(cacheDir, `package-${version}${path.extname(packageInfo.path) || ".7z"}`);
@@ -499,14 +498,14 @@ class AppUpdater extends events_1.EventEmitter {
             await downloadedUpdateHelper.clear().catch(() => {
                 // ignore
             });
-            return await promises_1.unlink(updateFile).catch(() => {
+            return await fs_extra_1.unlink(updateFile).catch(() => {
                 // ignore
             });
         };
         const tempUpdateFile = await DownloadedUpdateHelper_1.createTempUpdateFile(`temp-${updateFileName}`, cacheDir, log);
         try {
             await taskOptions.task(tempUpdateFile, downloadOptions, packageFile, removeFileIfAny);
-            await promises_1.rename(tempUpdateFile, updateFile);
+            await fs_extra_1.rename(tempUpdateFile, updateFile);
         }
         catch (e) {
             await removeFileIfAny();
```

This worked for me. Hope it helps.

**Comment by davej** [comment]
Thanks, @Mgonand. You can probably also just do a 1-line diff also.

```diff
-const promises_1 = require("fs/promises");
+const promises_1 = require("fs").promises;
```

**Comment by Mgonand** [comment]
You're right, that's better. I did the patch quick and didn't realize it. Thanks.

**Comment by axetroy** [comment]
`fs/promises` are great, but it breaking up electron version locking.

we need to upgrade the old version to electrion@^13

**Comment by toschlog** [comment]
Yes, I think this issue is still relevant.

I'm using electron-updater 4.3.9.  Differential downloads don't work.  Apparently they're fixed in 4.3.10, but that version won't work in Electron 10, which I have to use.

I'm going to try the patch-package thing but it would be good to release a 4.3.11 that fixes the breaking change in 4.3.10.

**Comment by mmaietta** [comment]
If I'm understanding this correctly, it seems we need to support node 14 (electron 12) and node 10 (electron 10). I wonder if we could use something like this for importing? Not at my dev machine to check
```
let fsPromises
try {
   fsPromises = require('fs/promises')
} catch (e} {
   fsPromises = require('fs').promises
}
```

**Comment by toschlog** [comment]
@mmaietta I ran your try/catch code and it worked fine.  Seems a good solution.

**Comment by mmaietta** [comment]
I noticed that `fs-extra` is already being used and the API's overlap, can someone try replacing this one import line to just swap in fs-extra like so?
```
const fs_extra_1 = require("fs-extra");
const promises_1 = fs_extra_1;
```
I'm working on getting this issue resolved so that you all are able to update your electron-updater to the newest alpha v5.

**Comment by mmaietta** [comment]
Migrated commands to use fs-extra and released this in `electron-updater@5.0.0-alpha.3` (note: alpha)
